### PR TITLE
Watermarks and linear aggregates.

### PR DESCRIPTION
**Pull request by ryzhyk**
This PR adds two operators required to implement nexmark q5(#146):
- time series watermark
- linear aggregates

### CODE REVIEW

File: src/operator/stream_fold.rs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    /// Folds every element in the input stream into an accumulator and outputs
+    /// the current value of the accumulator at every clock cycle.
+    ///
+    /// # Arguments
+    ///
+    /// * `init` - initial value of the accumulator.
+    /// * `fold_func` - closure that computes the new value of the accumulator
+    ///   as a function of the previous value and the new input at each clock
+    ///   cycle.
+    pub fn stream_fold<A, F>(&self, init: A, fold_func: F) -> Stream<Circuit<()>, A>
```
Comment (mbudiu-vmw) [inline src/operator/stream_fold.rs:21]: This seems to be a generalization of the integrator. For the integrator fold_func is just + and init is 0.

File: src/trace/layers/ordered.rs [hunk 2]
```diff
@@ -740,6 +740,19 @@ where
         }
     }
 
+    fn seek_last(&mut self) {
+        if self.bounds.1 > self.bounds.0 {
+            self.pos = self.bounds.1 - 1;
+
+            if self.valid() {
+                self.child.reposition(
```
Comment (mbudiu-vmw) [inline src/trace/layers/ordered.rs:748]: I can't figure out whether this is right, perhaps you can write a comment to justify it?

File: src/trace/layers/column_leaf/cursor.rs [hunk 3]
```diff
@@ -117,6 +117,13 @@ where
         self.seek_key(key.0);
     }
 
+    #[inline]
+    fn seek_last(&mut self) {
```
Comment (mbudiu-vmw) [inline src/trace/layers/column_leaf/cursor.rs:121]: again no indication whether this succeeded?

File: src/trace/mod.rs [hunk 4]
```diff
@@ -151,6 +151,8 @@ where
     /// Acquires a cursor to the batch's contents.
     fn cursor(&self) -> Self::Cursor<'_>;
 
+    /// The number of keys in the batch.
+    fn keys(&self) -> usize;
```
Comment (mbudiu-vmw) [inline src/trace/mod.rs:155]: how about key_count?

File: src/trace/ord/key_batch.rs [hunk 5]
```diff
@@ -365,6 +368,10 @@ where
         self.cursor.seek(key);
         self.valid = true;
     }
+    fn seek_last_key(&mut self) {
+        self.cursor.seek_last();
+        self.valid = true;
```
Comment (mbudiu-vmw) [inline src/trace/ord/key_batch.rs:373]: always?

**Review by mbudiu-vmw — APPROVED** [review]

File: src/operator/stream_fold.rs (hunk 1 above)
Comment (ryzhyk) [reply src/operator/stream_fold.rs:21]: True, but the integrator implementation is optimized for `+`, so it's not redundant.

File: src/trace/layers/column_leaf/cursor.rs (hunk 3 above)
Comment (ryzhyk) [reply src/trace/layers/column_leaf/cursor.rs:121]: the method always succeeds, but if the iterator is empty, `key_valid()` will be false.  This is inline with the rest of the cursor API.

File: src/trace/ord/key_batch.rs (hunk 5 above)
Comment (ryzhyk) [reply src/trace/ord/key_batch.rs:373]: Yes (see https://github.com/vmware/database-stream-processor/issues/145#issuecomment-1212317717)

File: src/trace/mod.rs (hunk 4 above)
Comment (ryzhyk) [reply src/trace/mod.rs:155]: ok, renamed to `key_count`

File: src/operator/aggregate.rs [hunk 6]
```diff
@@ -112,63 +107,104 @@ where
...
+            while cursor.key_valid() {
+                let mut agg = HasZero::zero();
+                while cursor.val_valid() {
+                    agg += f(cursor.key(), cursor.val()).mul_by_ref(&cursor.weight());
+                    cursor.step_val();
+                }
+                delta.push((cursor.key().clone(), agg));
+                cursor.step_key();
+            }
+            O::from_keys((), delta)
```
Comment (Kixiron) [inline src/operator/aggregate.rs:205]: Can't this use a builder since the key is unchanged?

File: src/trace/layers/ordered.rs (hunk 2 above)
Comment (Kixiron) [reply src/trace/layers/ordered.rs:750]: Some tests for the function would also be nice

**Review by Kixiron — APPROVED** [review]

File: src/operator/aggregate.rs (hunk 6 above)
Comment (ryzhyk) [reply src/operator/aggregate.rs:205]: Great idea!

**Pull request merged**